### ISSUE TITLE
Support tracing managed processes in gdb

### ISSUE THREAD

**Issue by sporksmith** [shadow/shadow#1873]
The `--gdb` flag is intended to help with this in general, but it pauses at a point before Shadow has started its managed processes, so it's not particularly useful for attaching to one of those.

We can trace shadow itself either by starting it under  gdb or attaching to it externally (with or without the `--gdb` flag), but tracing a managed process from its beginning is a little tricky. If we start from tracing shadow, we need to set `follow-fork-mode` to `child` to attach to the managed process, but if we set that from the beginning, then we'll follow the *first* fork, which is actually a utility process for getting CPU info.  

I've been working around this by setting a breakpoint at `_threadpreload_fork_exec` in shadow, and then changing `follow-fork-mode` to `child` just before the target process is started. It looks something like this:

```
gdb --args shadow ....
break _threadpreload_fork_exec
run
<check arguments to see which managed process is about to be forked>
set follow-fork-mode child  # if its the target process
continue
handle SIGSYS nostop
<set any other breakpoints etc>
continue
```

This isn't ideal as the supported way to do this though, since it involves setting a breakpoint on an internal function.

We could use something like `catch syscall vfork`. At least that would decouple us from the name of the internal function, but the user still needs to inspect the internals to see if the fork that's about to happen is going to exec the target process.

Something like the `--gdb` flag is probably a better solution, but I think we'll want to be able to specify a target process or simulation time to pause at.

**Comment by sporksmith** [comment]
We do have some documentation here: https://shadow.github.io/docs/guide/developer_guide.html#debugging-shadow-using-gdb. For managed processes it recommends using a core dump, but that's not always sufficient.

**Comment by stevenengler** [comment]
Todo:
- [ ] add a new cli option to help users debug specific managed processes
- support the workflows:
  - [ ] attach a new gdb session directly to the managed process
  - [ ] attach a new gdb session to shadow, and use the `follow-fork-mode` to switch from debugging shadow to debugging the managed process
- [ ] document the new option and the above workflows that use this option

**Comment by sporksmith** [comment]
Something we'll want to address and/or document:

On some systems - particularly Ubuntu - non-root users can only ptrace-attach to child processes of the current process. This is why `gdb -p` doesn't work on these systems without `sudo`. https://wiki.ubuntu.com/SecurityTeam/Roadmap/KernelHardening#ptrace_Protection

It looks like `CAP_SYS_PTRACE` rather than full root should also be adequate. This still requires access to the root user to get the capability in the first place, but at least lets you drop the rest of the root privileges and not disable the ptrace protection globally. `sudo capsh --caps=cap_sys_ptrace=ep --print -- -c "gdb -p 12129"` seems to work, but we should probably check a bit more carefully to get this right before documenting such a thing. Another option might be use Docker with the `CAP_SYS_PTRACE` capability enabled, but I haven't tried it.

I think we could work out an extension to support the `gdb --args shadow ...` use-case, where we pause just before the `fork`, at which point the user can use `set follow-fork-mode child` in gdb to switch over to the child that's about to be forked. That could be a later extension, though. I think we could even do it using the same config option. Since gdb stops on any signal, even if it's handled, we could install a no-op handler for e.g. SIGUSR1 and raise that signal just before forking processes in the given hosts. In gdb that'd cause it to pause. Outside of gdb it should be a no-op.

**Comment by stevenengler** [comment]
> attach a new gdb session to shadow, and use the follow-fork-mode to switch from debugging shadow to debugging the managed process

We decided not to support this use case using this `--debug-hosts` option since we haven't found a nice way to support both workflows.

If we still want this feature in the future, we can re-open this issue or open a new one.

> It would be nice if shadow checked to make sure that all of those hosts exist in the simulation, but this needs to be checked after hostnames are expanded due to the quantity option.

This was also not implemented since it'll be easier to do when the controller and manager are written in rust.

I think we've implemented everything else that we want. We have some documentation of this feature at https://shadow.github.io/docs/guide/debugging.html#attaching-with-gdb.
